[D3D9] The Void needs bits of IDirect3DShaderValidator9 implementation.

I've debugged 'The Void' Steam game some time ago (which was crashing at certain locations both with dxvk and Wine's d3d9) and figured that is the first known game which requires some bits of real IDirect3DShaderValidator9 implementation to work (I've put some details and PoC Wine implementation which adds the necessary bits to https://bugs.winehq.org/show_bug.cgi?id=47229).

The game disassembles and assembles back some shaders (I have no idea why). Some shaders fail to assemble back due to: 'dcl_texcoord10_pp v10' (only 10 inputs are supported in ps_3_0 and this is the eleventh) which result in the crash as the game doesn't consider D3DXAssembleShader() failures. With (partially) working shader validator the problematic shaders get rejected at the earlier stage (IIRC shader compilation from HLSL) and the game handles that fine.

The PoC patch for Wine's d3d9 attached to the Wine bug is enough to make game work.

For testing with the game, here is a save file and reproduction instructions: https://github.com/ValveSoftware/Proton/issues/4527. Also, here is the cheat sheet which explains how to open dev console and get to any location at once: https://forum.ice-pick.com/viewtopic.php?f=28&t=9044. Without that I would probably never get to a problematic spot.

Also, WRT game testing, it creates the compiled shader cache and once you have passed the problematic place successfully
the next run won't trigger problematic behaviour (e. g., if you first run on Windows without dxvk and then with dxvk you won't reproduce it until the cache is cleared). So it is better to delete users/Public/Documents/My\ Games/Void/Cache/* before each run.

There are a few errors that are easy to work around with:

Added "#include <utility>" to: ./src/dxvk/dxvk_buffer.h ./src/dxgi/dxgi_swapchain.cpp ./src/dxgi/dxgi_factory.cpp ./src/dxgi/dxgi_output.cpp ./src/dxgi/dxgi_main.cpp ./src/d3d10/d3d10_multithread.cpp

Added "#include <memory>" to: ./src/d3d10/d3d10_reflection.h

But then there's more complicated stuff like: 
```
../../../src/d3d9/d3d9_shader_validator.h:118:37: error: cannot bind non-const lvalue reference of type ‘dxvk::DxsoCodeIter&’ to an rvalue of type ‘dxvk::DxsoCodeIter’

118 |       if (!m_ctx->decodeInstruction(DxsoCodeIter{ reinterpret_cast<const uint32_t*>(pdwInst) }))
```

So I changed:
```
      DxsoReader reader = { reinterpret_cast<const char*>(pdwInst) };

      if (m_state == D3D9ShaderValidatorState::ValidatingHeader)
        return ValidateHeader(pFile, Line, pdwInst, cdw);

      if (!m_ctx->decodeInstruction(DxsoCodeIter{ reinterpret_cast<const uint32_t*>(pdwInst) }))
```
into (which might not be correct):
```
      DxsoReader reader = { reinterpret_cast<const char*>(pdwInst) };
      DxsoCodeIter iter(reinterpret_cast<const uint32_t*>(pdwInst));

      if (m_state == D3D9ShaderValidatorState::ValidatingHeader)
        return ValidateHeader(pFile, Line, pdwInst, cdw);

      if (!m_ctx->decodeInstruction(iter))
```

Which allowed me to build dxvk, unfortunately, the game still dies with this.

--- a/src/d3d9/d3d9_shader_validator.h
+++ b/src/d3d9/d3d9_shader_validator.h
@@ -2,22 +2,56 @@
 
 #include "d3d9_include.h"
 
+#include "../dxso/dxso_header.h"
+#include "../dxso/dxso_decoder.h"
+
 namespace dxvk {
+
+  enum class D3D9ShaderValidatorMessage : uint32_t {
+    BeginOutOfOrder = 0xeb,
+    InstructionOutOfOrder = 0xec,
+    InstructionEndOfShader = 0xed,
+    InstructionNullArgs = 0xee,
+    BadVersionTokenLength = 0xef,
+    BadVersionTokenType = 0xf0,
+    BadEndToken = 0xf1,
+    EndOutOfOrder = 0xf2,
+    MissingEndToken = 0xf3,
+    BadInputRegisterDeclaration = 0x12c,
+    BadInputRegister = 0x167,
+    BadInstructionLength = 0x21e,
+  };
+
+  enum class D3D9ShaderValidatorState {
+    Begin,
+    ValidatingHeader,
+    ValidatingInstructions,
+    EndOfShader,
+    Error,
+  };
+
+  using D3D9ShaderValidatorCallback = HRESULT(STDMETHODCALLTYPE *)(
+    const char*                      pFile,
+          UINT                       Line,
+          DWORD                      Unknown,
+          D3D9ShaderValidatorMessage MessageID,
+    const char*                      pMessage,
+          void*                      pUserData);
 
   class IDirect3DShaderValidator9 : public IUnknown {
 
   public:
 
     virtual HRESULT STDMETHODCALLTYPE Begin(
-            void* pCallback,
-            void* pUserParam,
-            DWORD Unknown) = 0;
+            D3D9ShaderValidatorCallback pCallback,
+            void*                       pUserParam,
+            DWORD                       Unknown) = 0;
 
     virtual HRESULT STDMETHODCALLTYPE Instruction(
-      const char*  pUnknown1,
-            UINT   Unknown2,
-      const DWORD* pInstruction,
-            DWORD  InstructionLength) = 0;
+      const char*  pFile,
+            UINT   Line,
+      const DWORD* pdwInst,
+            DWORD  cdw) = 0;
 
     virtual HRESULT STDMETHODCALLTYPE End() = 0;
 
@@ -37,32 +71,204 @@ namespace dxvk {
 
 
     HRESULT STDMETHODCALLTYPE Begin(
-            void* pCallback,
-            void* pUserParam,
-            DWORD Unknown) {
-      Logger::debug("D3D9ShaderValidator::Begin: Stub");
+        D3D9ShaderValidatorCallback pCallback,
+        void*                       pUserData,
+        DWORD                       Unknown) {
+      if (unlikely(m_state != D3D9ShaderValidatorState::Begin)) {
+        return ErrorCallback(nullptr, -1, 0,
+          D3D9ShaderValidatorMessage::BeginOutOfOrder,
+          "IDirect3DShaderValidator9::Begin called out of order. ::End must be called first.");
+      }
+
+      m_callback = pCallback;
+      m_userData = pUserData;
+      m_state    = D3D9ShaderValidatorState::ValidatingHeader;
 
       return D3D_OK;
     }
 
 
     HRESULT STDMETHODCALLTYPE Instruction(
-      const char*  pUnknown1,
-            UINT   Unknown2,
-      const DWORD* pInstruction,
-            DWORD  InstructionLength) {
-      Logger::debug("D3D9ShaderValidator::Instruction: Stub");
+        const char*  pFile,
+              UINT   Line,
+        const DWORD* pdwInst,
+              DWORD  cdw) {
+      if (unlikely(pdwInst == nullptr || !cdw)) {
+        return ErrorCallback(pFile, Line, 0,
+          D3D9ShaderValidatorMessage::InstructionNullArgs,
+          "IDirect3DShaderValidator9::Instruction called with NULL == pdwInst or 0 == cdw.");
+      }
+
+      if (unlikely(m_state == D3D9ShaderValidatorState::Begin)) {
+        return ErrorCallback(pFile, Line, 0,
+          D3D9ShaderValidatorMessage::InstructionOutOfOrder,
+          "IDirect3DShaderValidator9::Instruction called out of order. ::Begin must be called first.");
+      } else if (unlikely(m_state == D3D9ShaderValidatorState::EndOfShader)) {
+        return ErrorCallback(pFile, Line, 0,
+          D3D9ShaderValidatorMessage::InstructionEndOfShader,
+          "IDirect3DShaderValidator9::Instruction called out of order. After end token there should be no more instructions. Call ::End next.");
+      } else if (unlikely(m_state == D3D9ShaderValidatorState::Error)) {
+        return E_FAIL;
+      }
+
+      if (m_state == D3D9ShaderValidatorState::ValidatingHeader)
+        return ValidateHeader(pFile, Line, pdwInst, cdw);
+
+      DxsoCodeIter pdwInstIter{ reinterpret_cast<const uint32_t*>(pdwInst) };
+      bool isEndToken = !m_ctx->decodeInstruction(pdwInstIter);
+      const DxsoInstructionContext instContext = m_ctx->getInstructionContext();
+
+      if (isEndToken)
+        return ValidateEndToken(pFile, Line, pdwInst, cdw);
+
+      // TODO: DxsoDecodeContext::decodeInstructionLength() does not currently appear
+      // to return the correct token length in many cases, and as such dwordLength
+      // will not be equal to cdw in many situations that are expected to pass validation
+      //
+      /*Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: opcode ", instContext.instruction.opcode));
+      // + 1 to account for the opcode...
+      uint32_t dwordLength = instContext.instruction.tokenLength + 1;
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: cdw ", cdw));
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: dwordLength ", dwordLength));
+      if (unlikely(cdw != dwordLength)) {
+        return ErrorCallback(pFile, Line, 0x2,
+          D3D9ShaderValidatorMessage::BadInstructionLength,
+          str::format("Instruction length specified for instruction (", cdw, ") does not match the token count encountered (", dwordLength, "). Aborting validation."));
+      }*/
+
+      // a maximum of 10 inputs are supported with PS 3.0 (validation required by The Void)
+      if (m_isPixelShader && m_majorVersion == 3) {
+        switch (instContext.instruction.opcode) {
+          case DxsoOpcode::Comment:
+          case DxsoOpcode::Def:
+          case DxsoOpcode::DefB:
+          case DxsoOpcode::DefI:
+            break;
+
+          default:
+            // Iterate over register tokens. Bit 31 of register tokens is always 1.
+            for (uint32_t instNum = 1; pdwInst[instNum] & 0x80000000; instNum++) {
+              DWORD regType  = ((pdwInst[instNum] & D3DSP_REGTYPE_MASK)  >> D3DSP_REGTYPE_SHIFT)
+                             | ((pdwInst[instNum] & D3DSP_REGTYPE_MASK2) >> D3DSP_REGTYPE_SHIFT2);
+              DWORD regIndex = pdwInst[instNum] & D3DSP_REGNUM_MASK;
+
+              if (unlikely(regType == static_cast<DWORD>(DxsoRegisterType::Input) && regIndex >= 10)) {
+                return ErrorCallback(pFile, Line, 0x2,
+                  instContext.instruction.opcode == DxsoOpcode::Dcl ?
+                    D3D9ShaderValidatorMessage::BadInputRegisterDeclaration :
+                    D3D9ShaderValidatorMessage::BadInputRegister,
+                  "IDirect3DShaderValidator9::Instruction: Invalid number of PS input registers specified. Aborting validation.");
+              }
+            }
+        }
+      }
 
       return D3D_OK;
     }
 
 
     HRESULT STDMETHODCALLTYPE End() {
-      Logger::debug("D3D9ShaderValidator::End: Stub");
+      if (unlikely(m_state == D3D9ShaderValidatorState::Error)) {
+        return E_FAIL;
+      } else if (unlikely(m_state == D3D9ShaderValidatorState::Begin)) {
+        return ErrorCallback(nullptr, 0, 0,
+          D3D9ShaderValidatorMessage::EndOutOfOrder,
+          "IDirect3DShaderValidator9::End called out of order. Call to ::Begin, followed by calls to ::Instruction must occur first.");
+      } else if (unlikely(m_state != D3D9ShaderValidatorState::EndOfShader)) {
+        return ErrorCallback(nullptr, 0, 0,
+          D3D9ShaderValidatorMessage::MissingEndToken,
+          "IDirect3DShaderValidator9::End: Shader missing end token.");
+      }
+
+      m_state    = D3D9ShaderValidatorState::Begin;
+      m_isPixelShader = false;
+      m_majorVersion  = 0;
+      m_minorVersion  = 0;
+      m_callback = nullptr;
+      m_userData = nullptr;
+      m_ctx      = nullptr;
 
       return D3D_OK;
     }
 
+  private:
+
+    HRESULT ValidateHeader(const char* pFile, UINT Line, const DWORD* pdwInst, DWORD cdw) {
+      if (unlikely(cdw != 1)) {
+        return ErrorCallback(pFile, Line, 0x6,
+          D3D9ShaderValidatorMessage::BadVersionTokenLength,
+          "IDirect3DShaderValidator9::Instruction: Bad version token. DWORD count > 1 given. Expected DWORD count to be 1 for version token.");
+      }
+
+      DxsoReader reader = { reinterpret_cast<const char*>(pdwInst) };
+      uint32_t headerToken = reader.readu32();
+      uint32_t shaderType  = headerToken & 0xffff0000;
+      DxsoProgramType programType;
+
+      if (shaderType == 0xffff0000) { // Pixel Shader
+        programType = DxsoProgramTypes::PixelShader;
+        m_isPixelShader = true;
+      } else if (shaderType == 0xfffe0000) { // Vertex Shader
+        programType = DxsoProgramTypes::VertexShader;
+        m_isPixelShader = false;
+      } else {
+        return ErrorCallback(pFile, Line, 0x6,
+          D3D9ShaderValidatorMessage::BadVersionTokenType,
+          "IDirect3DShaderValidator9::Instruction: Bad version token. It indicates neither a pixel shader nor a vertex shader.");
+      }
+
+      m_majorVersion = (headerToken >> 8) & 0xff;
+      m_minorVersion = headerToken & 0xff;
+      m_ctx   = std::make_unique<DxsoDecodeContext>(DxsoProgramInfo{ programType, m_minorVersion, m_majorVersion });
+      m_state = D3D9ShaderValidatorState::ValidatingInstructions;
+
+      const char* shaderTypeOutput = m_isPixelShader ? "PS" : "VS";
+      Logger::debug(str::format("IDirect3DShaderValidator9::Instruction: Validating ",
+                                shaderTypeOutput, " version ", m_majorVersion, ".", m_minorVersion));
+
+      return D3D_OK;
+    }
+
+    HRESULT ValidateEndToken(const char* pFile, UINT Line, const DWORD* pdwInst, DWORD cdw) {
+      // Reached the end token.
+      if (unlikely(cdw != 1)) {
+        return ErrorCallback(pFile, Line, 0x6,
+          D3D9ShaderValidatorMessage::BadEndToken,
+          "IDirect3DShaderValidator9::Instruction: Bad end token. DWORD count > 1 given. Expected DWORD count to be 1 for end token.");
+      }
+
+      m_state = D3D9ShaderValidatorState::EndOfShader;
+
+      return D3D_OK;
+    }
+
+    HRESULT ErrorCallback(
+        const char*                      pFile,
+              UINT                       Line,
+              DWORD                      Unknown,
+              D3D9ShaderValidatorMessage MessageID,
+        const std::string&               Message) {
+      if (m_callback)
+          m_callback(pFile, Line, Unknown, MessageID, Message.c_str(), m_userData);
+
+      // TODO: Consider switching this to debug, once we're
+      // confident the implementation doesn't cause any issues
+      Logger::warn(Message);
+
+      m_state = D3D9ShaderValidatorState::Error;
+
+      return E_FAIL;
+    }
+
+    bool                        m_isPixelShader = false;
+    uint32_t                    m_majorVersion  = 0;
+    uint32_t                    m_minorVersion  = 0;
+
+    D3D9ShaderValidatorState    m_state         = D3D9ShaderValidatorState::Begin;
+    D3D9ShaderValidatorCallback m_callback      = nullptr;
+    void*                       m_userData      = nullptr;
+
+    std::unique_ptr<DxsoDecodeContext> m_ctx;
   };
 
 }

